registry/handlers/referrers.go

```diff
@@ -0,0 +1,364 @@
+					if err != nil {
+						return fmt.Errorf("failed to parse created annotation timestamp: %v", err)
+					}
+					referrersWrappers = append(referrersWrappers, referrersSortedWrapper{
+						createdAt:  extractedTimestamp,
+						descriptor: artifactDesc,
+					})
+				}
+			}
+			return nil
+		})
+
+	if err != nil {
+		switch err.(type) {
+		case driver.PathNotFoundError:
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	// sort the list of descriptors that contain the created annotation
+	sort.Slice(referrersWrappers, func(i, j int) bool {
+		// most recent artifact first
+		return referrersWrappers[i].createdAt.After(referrersWrappers[j].createdAt)
+	})
+	// extract the artifact descriptor from the sorting wrapper
+	for _, wrapper := range referrersWrappers {
+		referrersSorted = append(referrersSorted, wrapper.descriptor)
+	}
+	// append the descriptors, which don't have a created annotation, to the end
+	referrersSorted = append(referrersSorted, referrersUnsorted...)
+	return referrersSorted, nil
+}
+
+func enumerateReferrerLinks(ctx context.Context,
+	rootPath string,
+	stDriver driver.StorageDriver,
+	repository distribution.Repository,
+	blobstatter distribution.BlobStatter,
+	subjectRevision digest.Digest,
+	artifactManifestIndex map[digest.Digest][]digest.Digest,
+	ingestor func(ctx context.Context,
+		digest digest.Digest,
+		subjectRevision digest.Digest,
+		artifactManifestIndex map[digest.Digest][]digest.Digest,
+		repository distribution.Repository,
+		blobstatter distribution.BlobStatter,
+		storageDriver driver.StorageDriver) error) error {
+
+	return stDriver.Walk(ctx, rootPath, func(fileInfo driver.FileInfo) error {
+		// exit early if directory...
+		if fileInfo.IsDir() {
+			return nil
+		}
+		filePath := fileInfo.Path()
+
+		// check if it's a link
+		_, fileName := path.Split(filePath)
+		if fileName != "link" {
+			return nil
+		}
+
+		// read the digest found in link
+		digest, err := readlink(ctx, filePath, stDriver)
+		if err != nil {
+			return err
+		}
+
+		// ensure this conforms to the linkPathFns
+		_, err = blobstatter.Stat(ctx, digest)
+		if err != nil {
+			// we expect this error to occur so we move on
+			if err == distribution.ErrBlobUnknown {
+				return nil
+			}
+			return err
+		}
+
+		err = ingestor(ctx,
+			digest,
+			subjectRevision,
+			artifactManifestIndex,
+			repository,
+			blobstatter,
+			stDriver)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+func readlink(ctx context.Context, path string, stDriver driver.StorageDriver) (digest.Digest, error) {
+	content, err := stDriver.GetContent(ctx, path)
+	if err != nil {
+		return "", err
+	}
+
+	linked, err := digest.Parse(string(content))
+	if err != nil {
+		return "", err
+	}
+
+	return linked, nil
+}
+
+func generateLinkHeader(repoName, subjectDigest, artifactType string, lastDigests []string, nPage int) string {
+	url := fmt.Sprintf("/v2/%s/_referrers/manifests/referrers?digest=%s&nextToken=%s",
```

Comment:
@shizhMSFT Is this url okay? For this feature to work, this url can be anything. But I'm not sure if it has any semantic meaning.

Reply:
The URL is defined in https://github.com/opencontainers/distribution-spec/blob/main/spec.md#listing-referrers

There is no such a field called `nextToken`.

Reply:
Therefore, `nextToken` is an implementation specific field. You might want to document this somewhere.

Reply:
I think we may not need the `nextToken` field. I've created an issue to keep track of the paging of referrers, and I'll look deeper into it when fixing that issue.

Reply:
Updated the url according to the distribution spec.